Scoped Registry Framework

Zidane opened the pull request:
Needs some heavy documentation to explain the methods but assume any method that doesn't return an Optional will throw an exception should the value not exist.

Feel free to wait to review this until that exists but much of this is likely self explanatory.

To help with demonstration, `MatterTypes` has been updated.

Zidane committed:
Scoped registry framework.

As Minecraft continues to be more data-driven, the API must evolve and
serve as a guide to plugin developers. For game content, the engine is a
croassroads where 1.16.4 shows the start of game objects driven outside
code and clients receiving game content during connection.

This API hopes to future proof against such endeavours. Most importantly,
the API removes the concept of catalogs being singletons and treats
everything as references.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Allow creating a RegistryKey using another RegistryKey as the parent.

This effectively passes up the location and limits the chance of someone fudging the ids between the types and registries classes

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Holders should not be responsible for going through scopes.

That is the job of RegistryReference and most plugin devs will use those references regardless.

Might try and add some utility method somewhere, dunno where, for when you don't have a RegistryReference and want to go up scopes but we need to be careful that Vanilla may, in the future, decide to not check against indirect scopes should a direct scope not have a registry entry. Food for thought.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
More QOL, directly use a RegistryKey when querying a Registry.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Final look and feel for the types classes.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

dualspiral commented on src/main/java/org/spongepowered/api/registry/RegistryEntry.java:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @param <V> The type of the registry
+ */
+public interface RegistryEntry<V> {
```
I think this is going to be the key change here and we need to be totally clear about it.

Is this a replacement for the `CatalogType` interface? Given that, in theory, objects could be in multiple registries with different keys, this makes total sense because an object can no longer be identified by key alone - it's a combination of key and registry (or two keys if you _really_ want), and with scoping, as the server/worlds cycle, so too will the registry entries.

I think, in this case, this should contain the `RegistryKey`.

Zidane replied:
Negative, CatalogType isn't changing. This is simply a useful Key + Value struct for `Registry`. Could actually be tossed but would make `stream` on `Registry` annoying to use.

dualspiral replied:
Okay, got it.

dualspiral commented on src/main/java/org/spongepowered/api/registry/RegistryEntry.java:
```diff
@@ -0,0 +1,49 @@
+     *
+     * @return The key
+     */
+    ResourceKey key();
```
`RegistryKey`? Could have this too, but this isn't sufficient for identification purposes.

Zidane replied:
`RegistryKey` could be done here but if you have an entry, you at some point had/know of the actual `Registry` it came from.

Plugin devs are not expected to ever build entries.

dualspiral commented on src/main/java/org/spongepowered/api/event/lifecycle/RegisterRegistryValueEvent.java:
```diff
@@ -24,18 +24,18 @@
+import org.spongepowered.api.registry.ScopedRegistryHolder;
 
-public interface RegisterCatalogEvent<T extends CatalogType> extends GenericEvent<T>, LifecycleEvent {
+public interface RegisterRegistryValueEvent<H extends ScopedRegistryHolder, T> extends GenericEvent<T>, LifecycleEvent {
```
Are we assuming that every registry holder can only have one registry of type `T`?

There is also going to need to be another form of `GenericEvent` that accepts two generic parameters - or we turn this into three events, one for each scope.

Zidane replied:
No, in fact it is entirely possible for a `RegistryHolder` to have multiple registries of type `T`. This is why this event gives you the `RegistryKey` of the registry being invoked. Unfortunately, plugin devs will need to check against the API `Registries` constants if they want to register against a built-in registry. I don't see any clean away around it.

dualspiral replied:
No, neither do I. The only thing I can think of is adding a `Optional<RegistryEntry<T>> registerIf(RegistryKey<T> registryKey, ResourceKey key, T value)` if we want to allow plugin devs to be lazy about it - or perhaps make it the only register method to enforce devs to think about it.

There isn't a nice solution though, I agree.

dualspiral commented on src/main/java/org/spongepowered/api/item/recipe/RecipeRegistration.java:
```diff
@@ -26,9 +26,10 @@
 
 /**
- * A Sponge plugin recipe registration use {@link org.spongepowered.api.event.lifecycle.RegisterCatalogEvent<RecipeRegistration>} to register the recipe.
+ * A Sponge plugin recipe registration use {@link RegisterRegistryValueEvent <RecipeRegistration>} to register the recipe.
```
This has two type parameters, not one!

dualspiral commented on src/main/java/org/spongepowered/api/registry/Registry.java:
```diff
@@ -0,0 +1,91 @@
+     * @param value The value to register
+     * @return The added {@link RegistryEntry}
+     */
+    RegistryEntry<T> register(ResourceKey key, T value);
```
I'm not sure I necessarily agree with what I'm saying, but I'll say it anyway as an option: `Optional<RegistryEntry<T>> registerIfDynamic(ResourceKey key, T value);`? Returns empty if not dynamic - and a plugin can go on with its day.

Not wedded to that though.

Zidane replied:
Could gate it with an `Optional` and likely needs to be an `Optional` return regardless as there are other reasons failure could occur that won't necessitate an exception thrown.

dualspiral commented on src/main/java/org/spongepowered/api/registry/RegistryHolder.java:
```diff
@@ -0,0 +1,41 @@
+
+public interface RegistryHolder {
+
+    <R> Registry<R> registry(RegistryKey<R> key);
```
I guess this will throw an `IllegalArgumentException` on failure? (edit: and I should really read what you put in your initial comment before commenting... it will throw an exception)

Zidane replied:
These methods are really meant for the API use case and very confident plugin devs that know certain content is assured to exist.

That said, this new API is designed around the fact that you are guaranteed one thing and one thing alone: no value in a registry within a holder is ever guaranteed to exist. That will be the pain point of the new system but hopefully RegistryReference will help with this.

dualspiral commented on src/main/java/org/spongepowered/api/registry/RegistryScopes.java:
```diff
@@ -0,0 +1,37 @@
+@Target(ElementType.TYPE)
+public @interface RegistryScopes {
+
+    RegistryScope[] scopes();
```
`value()`

dualspiral reviewed (commented):
Initial thoughts within.

(General thought: This is going to be paaaainnnnffffuuullll.)

Zidane committed:
Documentation and class name adjustments.

I also expanded the registration events but need some input on this.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Fixup RecipeRegistration docs.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Many more registry changes, almost done with API structure.

- CatalogRegistry is no more
- Allow Registry getters to duck subtypes
- Rename Builder/Factory interfaces to providers (as they provide these objects)

This commit is the start of breaking EVERYTHING

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Update for Builder/Factory provider changes.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Fix API issues with registries.

The registry system is built around having "roots" who have registries underneath. A pseuo "registries of registries". Unless it is a real Minecraft registry, all other registries will exist under Sponge's root.

- RegistryKey's registry cannot be a ResourceKey nor itself as it loses the root with the latter and ends up in infinite recursion with the former leading tooo
- Add RegistryLocation which nets the registry location itself and it's root. In the API, there are 2 distinct roots: minecraft and sponge
- Fix the generics on Registries as these are RegistryKeys of registries themselves
- Add RegistryRoots as a way to get the roots provided by the API so that keys off of roots can be constructed

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Fixup these classes to follow code standards.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
More type fixes for standards.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Whoops

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Bah

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Sigh...

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Even more sighs...

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
First step, all registries.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Add the two already converted registries.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Commit:
Registry name

Zidane committed:
Add DefaultedRegistryReference to fix an API usability issue created due to dropping get()

This adds maximum flexibility by allowing types such as BiomeTypes to not be defaulted but allow the other types to express a holder for the standard get()

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Fix generator mistake.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Update all types for references. API compiles.

kashike needs to review the types and apply the appropriate keys (ones that are Minecraft vs. Sponge)

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

gabizou commented:
Was manually merged in https://github.com/SpongePowered/SpongeAPI/commit/7ae0c2e79483511708130984a24861b3cf34bab3
